### ISSUE TITLE
Menubars: Tab wraps but Shift+Tab / Backspace do not

### ISSUE THREAD

**Issue by dannye** [fltk/fltk#1156]
**Describe the bug**

TL;DR: In menubars, Shift+Tab and Backspace behave inconsistently with Tab in 3 different ways.

From https://github.com/fltk/fltk/issues/1155#issuecomment-2526191563

1. In 1.4.0, using Tab to move down a submenu wraps back to the start, but Shift+Tab and Backspace do not wrap. (It might be worth noting that the up/down arrow keys do not wrap either, and this is probably intentional. Only left/right arrow keys wrap for topmenu items. I'm not suggesting any changes to arrow key handling, just drawing comparisons.)

2. Also, if Shift+Tab or Backspace is held so that the selection moves back to the start, then right (for topmenus) or down (for submenus) needs to be pressed twice in order to move the selection once.
(This is probably a mostly unimportant detail, because if the wrapping is fixed then this bug would probably go away. But just mentioning for completeness.)

3. If a topmenu item is currently highlighted, then Tab will move down and enter the first submenu item (it does _not_ move right to the next topmenu item). But Shift+Tab moves left to the previous topmenu item (whereas I would expect it to move to the last/bottom submenu item (like the actual opposite of Tab)).

**To Reproduce**

Bugs 1 and 2:

1. Run test/menubar
2. Click on a menu item in the middle, like Font
3. Press Tab
4. See that the selection moves down into the submenu
5. Hold Tab
6. See that the selection wraps within that submenu
7. Hold Shift+Tab or Backspace
8. Bug 1: See that the selection does not wrap (stops at the top of the submenu)
9. Bug 2: Notice that you must press down TWICE to move the selection

Bug 3:

1. Run test/menubar
2. Click on a menu item in the middle, like Font
3. Press Shift+Tab or Backspace
4. Bug 3: See that the selection moves left to Radio instead of the bottom of the Font submenu (the arrow symbol menu item below Shadow)
5. From here, pressing Tab goes down into the Radio submenu (which is expected, just pointing out the contrast with the Shift+Tab press from steps 3/4)

**Expected behavior**

Shift+Tab and Backspace should probably wrap like Tab.
Shift+Tab and Backspace should probably stick to submenus like Tab.

**Screenshots**

Screenshots/videos are probably more confusing than helpful in this case because you can't see what my keyboard is doing so it's unclear when I'm using tab/backspace/arrow keys.

**FLTK Version**
 - Version: 1.4.0
 - Downloaded and self-built from: Git
 - If from Git, branch: master
 - If from Git, commit: 5ff42dd

**FLTK Configure / Build Options**
 - Default CMake build

**Operating System / Platform:**
 - OS: Multiple
 
 **Additional context**

Bugs 1 and 2 can also be seen with any of the choice dropdowns in test/menubar.
Bug 3 is unique to top level menubars.

**Comment by dannye** [comment]
Bug 3 could also be fixed instead by changing Tab to cycle through the topmenu (ie, go right) instead of entering the submenu (go down). That one is pretty subjective.

Regardless, all of these just minor quirks, not hugely important..

**Comment by dannye** [comment]
The following change fixes Bugs 1 and 2:

```diff
diff --git a/src/Fl_Menu.cxx b/src/Fl_Menu.cxx
index 6332673da..715604092 100644
--- a/src/Fl_Menu.cxx
+++ b/src/Fl_Menu.cxx
@@ -798,7 +798,7 @@ int menuwindow::handle_part1(int e) {
     case FL_BackSpace:
     BACKTAB:
       if (!backward(pp.menu_number)) {
-        pp.item_number = -1;
+        pp.item_number = pp.menu_number >= 0 ? pp.p[pp.menu_number]->numitems : -1;
         backward(pp.menu_number);
       }
       return 1;
```

**Comment by dannye** [comment]
And this change fixes Bug 3 (by changing Tab to act more like Shift+Tab for top level menubars):

```diff
diff --git a/src/Fl_Menu.cxx b/src/Fl_Menu.cxx
index 6332673da..9868a1eb6 100644
--- a/src/Fl_Menu.cxx
+++ b/src/Fl_Menu.cxx
@@ -711,7 +711,7 @@ static int forward(int menu) { // go to next item in menu menu if possible
     }
     item = -1;
   }
-  while (pp.menubar && Fl::event_key() == FL_Right);
+  while (pp.menubar && (Fl::event_key() == FL_Right || Fl::event_key() == FL_Tab));
   return 0;
 }
 
@@ -813,6 +813,7 @@ int menuwindow::handle_part1(int e) {
       return 1;
     case FL_Tab:
       if (Fl::event_shift()) goto BACKTAB;
+      if (pp.menubar && pp.menu_number == 0) goto RIGHT;
     case FL_Down:
       if (pp.menu_number || !pp.menubar) {
         if (!forward(pp.menu_number) && Fl::event_key()==FL_Tab) {
```

**Comment by dannye** [comment]
And this fixes the issue where Shift+Tab / Backspace are ignored for top level menubars if the mouse is moved off of the menu:

```diff
diff --git a/src/Fl_Menu.cxx b/src/Fl_Menu.cxx
index 6332673da..69a8e30d3 100644
--- a/src/Fl_Menu.cxx
+++ b/src/Fl_Menu.cxx
@@ -797,7 +797,9 @@ int menuwindow::handle_part1(int e) {
     switch (Fl::event_key()) {
     case FL_BackSpace:
     BACKTAB:
-      if (!backward(pp.menu_number)) {
+      if (pp.menubar && pp.menu_number<0)
+        backward(0);
+      else if (!backward(pp.menu_number)) {
         pp.item_number = -1;
         backward(pp.menu_number);
       }
```
